SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialRPCService.cpp

```diff
@@ -0,0 +1,430 @@
+
+	EntityComponentId EntityComponent = EntityComponentId(EntityId, RingBufferComponentId);
+	EntityRPCType EntityType = EntityRPCType(EntityId, Type);
+
+	Schema_Object* EndpointObject;
+	uint64 LastAckedRPCId;
+	if (View->HasComponent(EntityId, RingBufferComponentId))
+	{
+		if (!View->HasAuthority(EntityId, RingBufferComponentId))
+		{
+			return EPushRPCResult::NoRingBufferAuthority;
+		}
+
+		Schema_ComponentUpdate** ComponentUpdatePtr = PendingComponentUpdatesToSend.Find(EntityComponent);
+		if (ComponentUpdatePtr == nullptr)
+		{
+			ComponentUpdatePtr = &PendingComponentUpdatesToSend.Add(EntityComponent, Schema_CreateComponentUpdate());
+		}
+		EndpointObject = Schema_GetComponentUpdateFields(*ComponentUpdatePtr);
+
+		if (Type == ERPCType::NetMulticast)
+		{
+			// Assume all multicast RPCs are auto-acked.
+			LastAckedRPCId = LastSentRPCIds[EntityType];
+		}
+		else
+		{
+			// We shouldn't have authority over the component that has the acks.
+			if (View->HasAuthority(EntityId, RPCRingBufferUtils::GetAckComponentId(Type)))
+			{
+				return EPushRPCResult::HasAckAuthority;
+			}
+
+			LastAckedRPCId = GetAckFromView(EntityId, Type);
+		}
+	}
+	else
+	{
+		// If the entity isn't in the view, we assume this RPC was called before
+		// CreateEntityRequest, so we put it into a component data object.
+		Schema_ComponentData** ComponentDataPtr = PendingRPCsOnEntityCreation.Find(EntityComponent);
+		if (ComponentDataPtr == nullptr)
+		{
+			ComponentDataPtr = &PendingRPCsOnEntityCreation.Add(EntityComponent, Schema_CreateComponentData());
+			LastSentRPCIds.Add(EntityType, 0);
+		}
+		EndpointObject = Schema_GetComponentDataFields(*ComponentDataPtr);
+
+		LastAckedRPCId = 0;
+	}
+
+	uint64 NewRPCId = LastSentRPCIds[EntityType] + 1;
+
+	// Check capacity.
+	if (LastAckedRPCId + RPCRingBufferUtils::GetRingBufferSize(Type) >= NewRPCId)
+	{
+		RPCRingBufferUtils::WriteRPCToSchema(EndpointObject, Type, NewRPCId, MoveTemp(Payload));
+
+		LastSentRPCIds[EntityType] = NewRPCId;
+	}
+	else
+	{
+		// Overflowed
+		if (RPCRingBufferUtils::ShouldQueueOverflowed(Type))
+		{
+			return EPushRPCResult::QueueOverflowed;
+		}
+		else
+		{
+			return EPushRPCResult::DropOverflowed;
+		}
+	}
+
+	return EPushRPCResult::Success;
+}
+
+EPushRPCResult SpatialRPCService::PushOverflowedRPCs()
+{
+	bool bOverflowedRPCsRemaining = false;
+
+	for (auto It = OverflowedRPCs.CreateIterator(); It; ++It)
+	{
+		Worker_EntityId EntityId = It.Key().EntityId;
+		ERPCType Type = It.Key().Type;
+		TArray<RPCPayload>& OverflowedRPCArray = It.Value();
+
+		int NumProcessed = 0;
+		for (RPCPayload& Payload : OverflowedRPCArray)
+		{
+			EPushRPCResult Result = PushRPCInternal(EntityId, Type, Payload);
+			if (Result == EPushRPCResult::Success)
+			{
+				NumProcessed++;
+			}
+			else if (Result == EPushRPCResult::QueueOverflowed)
+			{
+				bOverflowedRPCsRemaining = true;
+				break;
+			}
+			else
+			{
+				// Shouldn't happen, let the caller log.
+				return Result;
```

Comment:
We might want to put an error log here and drop everything as this means some invariant is broken.